Book - Handling Input Chapter & Initial Page (#2417)
bevyengine/bevy-website: content/learn/book/handling-input/using-input.md

+++
title = "Using Input"
insert_anchor_links = "right"
[extra]
weight = 1
+++

The process for reading and interacting with input data will generally be the same across all devices that Bevy can use.
When the button on a gamepad or a key on a keyboard gets pressed, Bevy uses [Winit] (via [`bevy_winit`], Bevy's conversion layer for Winit) to initially turn the input into a [`Message`].
These `Message`s are then processed and placed into a [`ButtonInput`] resource (or a `ButtonInput` component for gamepads) that we can then read and use for setting up movement, tracking aim, activating abilities, or any other input-based actions you'd like to set up.

[Winit]: https://crates.io/crates/winit
[`bevy_winit`]: https://docs.rs/bevy/latest/bevy/winit/index.html

> ## Using ButtonInput Versus Input Messages

Reviewer: This feels like a "patterns" bit.

+++
title = "Using Input"
insert_anchor_links = "right"
[extra]
weight = 1
+++

> Bevy can read input data from gamepads (which includes controllers), keyboards, mice, and touch inputs.

Reviewer 2: Gamepads are a subset of controllers, not the other was around. Do we support non-gamepad controllers?

Author: Not officially I don't think, but I'm sure you'd be able to make your own support if you dug into winit. But that's definitely outside the scope of the book.

The process for reading and interacting with input data will generally be the same across all devices.

> When the button on a gamepad or a key on a keyboard gets pressed, Bevy uses [Winit] (via [`bevy_winit`], Bevy's conversion tool for Winit) to initially turn the input into a [`Message`].

Reviewer 2: ```suggestion
When the button on a gamepad or a key on a keyboard gets pressed, Bevy uses [Winit] (via [`bevy_winit`], Bevy's conversion layer for Winit) to initially turn the input into a [`Message`].
```
Tool kinda makes me think of something used during development? Layer might be a bit clearer here.

> These `Messages` are then processed and placed into a resource (or a component) that we can then read and use for setting up movement, tracking aim, activating abilities, or any other input-based actions you'd like to set up.

Reviewer 2: ```suggestion
These `Message`s are then processed and placed into a resource (or a component) that we can then read and use for setting up movement, tracking aim, activating abilities, or any other input-based actions you'd like to set up.
```

> Before we cover the general process though, you should know that each device type also has their own unique circumstances to be aware of.

Reviewer 2: ```suggestion
```
Doesn't add anything not said in the next sentence.

While input data from every device will follow the same process that is covered on this page, you should read each devices' page to see how their data is uniquely handled:

- [Keyboard Input](/learn/book/handling-input/keyboard-input)
- [Mouse Input](/learn/book/handling-input/mouse-input)
- [Gamepad Input](/learn/book/handling-input/gamepad-input)
- [Touch Input](/learn/book/handling-input/touch-input)

When you use larger groupings of features (like profiles and collections), all of these devices are enabled in your game by default.
However we can adjust which ones are enabled by manually enabling their [feature flag].

> If you know that your game will not need touch input (or keyboard and mouse inputs if you're building a mobile game), you can disable these input devices by turning their feature flag off in your project `Cargo.toml` file.

Reviewer 2: Maybe leave that out – afaik you can only disable keyboard/mouse/touch features if you don't use `bevy_window`.

Author: I think it is worth mentioning since you can explicitly enabled the `keyboard` and `mouse` feature flags (and thus disable `touch` and `gamepad` inputs). Before we had the top level profiles, I went through and individually enabled the features I wanted in my projects, and I do know that I specifically did not enable `gamepad` or `touch` input. 

Granted, I think I could probably re-word this to shift from "disabling" (which isn't really possible with cargo) to "specifically enabling", which might be better for then pointing the reader at the relevant book page.

See the [Selective Feature Use section] on the Compiling Less Code page for more details.

[Selective Feature Use section]: /learn/book/releasing-projects/compiling-less-code/#more-selective-feature-use

[Winit]: https://crates.io/crates/winit
[`bevy_winit`]: https://docs.rs/bevy/latest/bevy/winit/index.html
[feature flag]: https://docs.rs/bevy/latest/bevy/index.html#feature-list

## Input Messages

Bevy takes input data from a device and converts it into a [`Message`].
These messages can be [read like any other message], and will have a unique type for each device: [`KeyboardInput`] for keyboards, [`MouseButtonInput`] for mouse button presses, and so on for each input type.
Each unique `Message` type will be automatically set up for the enabled devices in your game.
Bevy will also insert systems in the [`PreUpdate`] schedule to process, store, and eventually clear each `Message` type.

{% callout(type="info") %}

> ### Using Input Messages

Reviewer 2: I think this callout should just be part of the introduction of this page, since it's about when to use input `Messages` and then to use `ButtonInput` (this is kinda already indicated by the heading not really fitting – the whole "Input Messages" section *except* this callout is about "Using Input Messages").

While receiving input messages is the way that Bevy accesses input data, this doesn't mean that you should use input messages in every scenario.

> Input messages are only sent when an input is initially activated and then re-sent periodically if the input is still being activated (i.e. if a button is being pressed down or a joystick is continuously being pushed in a direction).

Reviewer 2: Holding down a mouse button does not repeatedly send messages.

These types of mechanics should instead be accessing the [`ButtonInput` resources] that we'll cover in the next section.

Instead, input messages are best suited for testing and logging input events, tracking text input, and activating systems or logic that don't rely on consistently repeated input.

Additionally, if you do choose to use input messages for some functionality, be wary of the `repeat` field on an input message which indicates that the input is being repeated (such as when a key or button is held down).
If you do not have some control for dealing with the `repeat` field, your functionality might cancel itself out, or flicker multiple times until `repeat` eventually returns `false`.

[`ButtonInput` resources]: /learn/book/handling-input/using-input#buttoninput-resources

{% end %}

Accessing input data through messages gives us access to all of the regular functionality that messages provide, including [`MessageReader`], [`MessageWriter`], and [`MessageMutator`].

```rust
// This system reads and prints out all `KeyboardInput` messages.
fn keyboard_message_reader(keyboard_inputs: MessageReader<KeyboardInput>) {
    for keyboard_input in keyboard_inputs.read() {

>         info!("{:?}", keyboard_input);

Reviewer 2: ```suggestion
        info!("{keyboard_input:?}");
```

    }
}
// This system writes a new `KeyboardInput` message.
fn keyboard_message_writer(
    mut keyboard_inputs: MessageWriter<KeyboardInput>,

>     window_query: Single<(Entity, &Window)>,

Reviewer 2: ```suggestion
    window_query: Single<Entity, With<Window>>,
```

- [Keyboard Input](/learn/book/handling-input/keyboard-input)
- [Mouse Input](/learn/book/handling-input/mouse-input)
- [Gamepad Input](/learn/book/handling-input/gamepad-input)
- [Touch Input](/learn/book/handling-input/touch-input)

When you use larger groupings of features (like profiles and collections), all of these devices are enabled in your game by default.
However we can adjust which ones are enabled by manually enabling their [feature flag].
If you know that your game will not need touch input (or keyboard and mouse inputs if you're building a mobile game), you can disable these input devices by turning their feature flag off in your project `Cargo.toml` file.
See the [Selective Feature Use section] in the Compiling Less Code page for more details.

[Selective Feature Use section]: /learn/book/releasing-projects/compiling-less-code/#more-selective-feature-use

[Winit]: https://crates.io/crates/winit
[`bevy_winit`]: https://docs.rs/bevy/latest/bevy/winit/index.html
[feature flag]: https://docs.rs/bevy/latest/bevy/index.html#feature-list

> ## Input Messages

Reviewer 3: I think this whole section should be lower down - putting it as the first makes it seem more "important" than the `ButtonInput` resources section, even though these messages aren't often used directly, and can cause issues if you forget to check for `repeat`
It should also clarify that the `ButtonInput` resource is usually used for game input, and this is meant more for things like text input (since that's when you _do_ want to respect the repeat events)

Since crouching might have a slight delay before being fully active, we'll want to make sure that we're constantly moving first rather than crouching.
This ensures that the player won't be interrupted by crouching while moving.

```rust
#[derive(Component)]
enum CrouchState {
    Uncrouched,
    Crouched
}

fn movement_plugin(mut app: &mut App) {
    // Using `.chain()` allows us to ensure that WASD movements get applied before Crouch movements.
    app.add_systems(FixedUpdate, (wasd_movement, crouching_system).chain());
}

fn wasd_movement(
    movement_input: MessageReader<KeyboardInput>,
    mut player_location: Single<&mut Transform, With<Player>>,
) {
    for input in movement_input.read() {
        // Match the input message based on what keyboard key is being pressed.
        match input.key_code {
            KeyCode::W => player_location.0.translation.y += 0.1,
            KeyCode::A => player_location.0.translation.x -= 0.1,
            KeyCode::S => player_location.0.translation.y -= 0.1,
            KeyCode::D => player_location.0.translation.x += 0.1,
            _ => continue;
        }
    }

> }

Reviewer 3: that's not how you should use the `KeyboardInput` messages, they're not emitted every frame. They're emitted when the key is first pressed, then every so often if you keep it held after a small delay - that's not what you want for game movement

[`MouseButtonInput`]: https://docs.rs/bevy/latest/bevy/input/mouse/struct.MouseButtonInput.html
[`MessageReader`]: https://docs.rs/bevy/latest/bevy/prelude/struct.MessageReader.html
[`MessageWriter`]: https://docs.rs/bevy/latest/bevy/prelude/struct.MessageWriter.html
[`MessageMutator`]: https://docs.rs/bevy/latest/bevy/ecs/message/struct.MessageMutator.html

## ButtonInput Resources

Once most "button-like" input messages are processed, they're placed into an associated [`ButtonInput`] resource that we can directly access via a system parameter.
Input data is added to a `ButtonInput` via the same systems that process and clear the input `Message` data.
These same systems will also clear each registered `ButtonInput` resource every frame.
However, when `ButtonInput` is cleared, [change detection] is not triggered.

{% callout(type="info") %}

### What Makes An Input "Button-like"?

When you start looking at each input device, you might wonder what makes an input "button-like"?

It's actually very straightforward.
To be considered "button-like" in Bevy, the input has to be "press-able".
This means that Bevy can register the state of the input as either `pressed` or `released`.

> Both of these values are explicitly stored in a [`ButtonState`] enum, which is apart of every "button-like" input `Message` that is sent.

Reviewer 3: ```suggestion
Both of these values are explicitly stored in a [`ButtonState`] enum, which is a part of every "button-like" input `Message` that is sent.
```
nit: a part means part of, apart means separated from

    player_weapon: Single<(&Weapon, &WeaponAttackInterval), With<PlayerWeapon>>
    time: Res<Time>,
) {
    let delta_time = time.delta();
    
    // Progress the WeaponAttackInterval.
    player_weapon.1.tick(delta_time);
    
    // If LeftMouseButton is pressed and our WeaponAttackInterval has completed, attack.
    if button_input.pressed(MouseButton::LeftMouseButton) && player_weapon.1.just_finished() {
        player_weapon.0.attack();
    }
}
```

However, we don't want the cooldown timer within the `WeaponAttackInterval` component to be running arbitrarily.
We only want it to run after the player has made an initial attack.
To start the timer in `WeaponAttackInterval`, we can now use `just_pressed` to start the cooldown timer after an initial attack is made.

```rust
fn weapon_attack(
    button_input: Res<ButtonInput<MouseButton>>,
    mut player_weapon: Single<(&Weapon, &mut WeaponAttackInterval), With<PlayerWeapon>>
) {
    if button_input.just_pressed(MouseButton::LeftMouseButton) {
        // Perform an initial weapon attack.
        player_weapon.0.attack();
        // Create a new the timer within `WeaponAttackInterval` that will run.

>         player_weapon.1.from_seconds(1.5, TimerMode::Repeating);

Reviewer 3: ```suggestion
        player_weapon.1.0 = Timer::from_seconds(1.5, TimerMode::Repeating);
```
nit: `Timer::from_seconds` isn't an instance method, it's a constructor

    }
}
```

Finally, we can extend this mechanic even further by using `just_released` to change the `WeaponAttackInterval` timer mode to `Once` when the player finally releases the attack button.
This will still let the timer tick down and finish.

```rust
fn cancel_weapon_attack_timer(
    button_input: Res<ButtonInput<MouseButton>>,
    mut player_weapon: Single<&mut WeaponAttackInterval, With<PlayerWeapon>>
) {
    if button_input.just_released(MouseButton::LeftMouseButton) {
        player_weapon.0.set_mode(TimerMode::Once);
    }
}
```

Since all three of these systems are interacting with the same data, we can combine them into one single `weapon_attack` system that will pick what happens when mouse input is received.

```rust
fn weapon_attack(
    button_input: Res<ButtonInput<MouseButton>>,
    mut player_weapon: Single<(&Weapon, &mut WeaponAttackInterval), With<PlayerWeapon>>,
    time: Res<Time>,
) {
    let delta_time = time.delta();
    
    // Check the state of the WeaponAttackInterval timer if it's active.

>     if player_weapon.1.is_finished() != true {

Reviewer 3: ```suggestion
    if !player_weapon.1.is_finished() {
```
nit: use `!x` instead of `x != true`

        // Progress the WeaponAttackInterval.
        player_weapon.1.tick(delta_time);
    }
    
    // The initial LeftMouseButton press.
    if button_input.just_pressed(MouseButton::LeftMouseButton) {
        // Perform an initial weapon attack.
        player_weapon.0.attack();
        // Create a new the timer within `WeaponAttackInterval` that will run.

>         player_weapon.1.from_seconds(1.5, TimerMode::Repeating);

Reviewer 3: ```suggestion
        player_weapon.1.0 = Timer::from_seconds(1.5, TimerMode::Repeating);
```
nit: constructor not instance method

    }
    
    // If LeftMouseButton is pressed and our WeaponAttackInterval has completed, attack.
    if button_input.pressed(MouseButton::LeftMouseButton) && player_weapon.1.just_finished() {
        player_weapon.0.attack();
    }
    
    // If LeftMouseButton was released, switch the timer mode to expire.
    if button_input.just_released(MouseButton::LeftMouseButton) {
        player_weapon.1.set_mode(TimerMode::Once);
    }
}
```

You might have noticed that this example only `ticks` (progresses) the timer when receiving input.
Since input data is only received each frame, if we experience lag or dropped frames, the weapon timer won't function consistently.
We won't cover a fix in this example since the intent is to provide you with an idea of how `pressed`, `just_pressed`, and `just_released` can be effectively used.

> If you'd like to see how we can fix this issue, check out the [Timers and Input] section on the Input Patterns page.

Reviewer 3: That's not true, these systems run every `Update`, so it ticks every update

+++
title = "Utilizing Input"

> insert_anchor_links = "right"

Reviewer: This page / related pages should talk about time + input: it's really important to get right.

fn register_system(mut commands: Commands) {
    let registered_system = commands.register_system(mouse_click_system);
}
```

[read like any other message]: /learn/book/control-flow/messages

[`Message`]: https://docs.rs/bevy/latest/bevy/ecs/message/trait.Message.html
[`PreUpdate`]: https://docs.rs/bevy/latest/bevy/app/struct.PreUpdate.html
[`KeyboardInput`]: https://docs.rs/bevy/latest/bevy/input/keyboard/struct.KeyboardInput.html
[`MouseButtonInput`]: https://docs.rs/bevy/latest/bevy/input/mouse/struct.MouseButtonInput.html
[`MessageReader`]: https://docs.rs/bevy/latest/bevy/prelude/struct.MessageReader.html
[`MessageWriter`]: https://docs.rs/bevy/latest/bevy/prelude/struct.MessageWriter.html
[`MessageMutator`]: https://docs.rs/bevy/latest/bevy/ecs/message/struct.MessageMutator.html

## ButtonInput Resources

Once most input messages are processed, they're placed into an associated [`ButtonInput`] resource that we can directly access via a system parameter.
Input data is added to a `ButtonInput` via the same systems that process and clear the input `Message` data, meaning we'll only have access to the input data as long as the original input `Message` is still available.

> Interacting with input data this way provides us with easier access to the state of the button input and gives us more control over how we respond to it. For example, accessing a [`ButtonInput`] resource provides us with a number of methods that will return a `bool` based on if the button [has just been pressed], [is currently being pressed], or [if its just been released].

Reviewer: The distinction between pressed / just pressed etc and when you should use each one deserves more words.

Instead, input messages are best suited for testing and logging input events, tracking text input, and activating systems or logic that don't rely on consistently repeated input.

> Additionally, if you do choose to use input messages for some functionality, be wary of the `repeat` field on an input message which indicates that the message will be re-sent.

Reviewer 3: the `repeat` field indicates that it's a repeated event (like when holding down a key), not that it will be resent

Reviewer 2: Additionally, it's only for keyboard input. Holding down a mouse button does not repeatedly fire events.

Reading input messages can be beneficial in several situations.
If we're dealing with multiple types of input data, input messages can allow us to easily order the input data relative to each other within a single frame.

As an example, let's setup a pause game mechanic.
Using input messages lets us iterate over each input message sent for the `KeyboardInput` message type.
If we detect that `KeyCode::Esc` is pressed, we'll evaluate whether the game is currently in `GameState::Playing` or `GameState::Paused`.
Based on this, we'll queue a transition to the opposite `GameState` variant.

```rust
fn ensure_pause(
    button_inputs: MessageReader<KeyboardInput>,
    current_game_state: Res<State<GameState>>,
    mut next_game_state: ResMut<NextState<GameState>>,
) {
    for input in button_inputs.iter() {

>         if input.key_code == KeyCode::Esc {

Reviewer 3: ```suggestion
        if input.key_code == KeyCode::Esc && !input.repeat {
```
You need to make sure that repeat events don't toggle the pause menu again. I think we should also explicitly warn about that somewhere (in the info box?)
